File: solutions/212.word-search-ii.js
/*
 * @lc app=leetcode id=212 lang=javascript
 *
 * [212] Word Search II
 */

// @lc code=start
/**
 * @param {character[][]} board
 * @param {string[]} words
 * @return {string[]}
 */
var findWords = function (board, words) {
  class Trie {
    insert(arr) {
      let node = this;
      for (const e of arr) {
        node[e] = node[e] ?? new Trie();
        node = node[e];
      }
      node.isEnd = true;
    }
    has(arr) {
      let node = this;
      for (const e of arr) {
        if (!(e in node)) return false;
        node = node[e];
      }
      return node.isEnd;
    }
    startsWith(prefixArr) {
      let node = this;
      for (const e of prefixArr) {
        if (!(e in node)) return false;
        node = node[e];
      }
      return true;
    }
  }

  const trie = new Trie();
  words.forEach((word) => trie.insert(word.split("")));
  const m = board.length;
  const n = board[0].length;
  const visited = Array.from({ length: m }, () =>
    Array.from({ length: n }, () => false)
  );
  const res = new Set();
  const dfs = (i, j, curStr) => {
    if (i < 0 || j < 0 || i >= m || j >= n || visited[i][j]) return;
    curStr += board[i][j];
    if (!trie.startsWith(curStr.split(""))) return;
    if (trie.has(curStr.split(""))) res.add(curStr);

    visited[i][j] = true;
    dfs(i + 1, j, curStr);
    dfs(i - 1, j, curStr);
    dfs(i, j + 1, curStr);
    dfs(i, j - 1, curStr);
    visited[i][j] = false;
  };

  for (let i = 0; i < m; ++i) {
    for (let j = 0; j < n; ++j) {
      dfs(i, j, "");
    }
  }
  return [...res];
};
// @lc code=end

File: solutions/212.word-search-ii.rb
# @param {Character[][]} board
# @param {String[]} words
# @return {String[]}
def find_words(board, words)
  m = board.size
  n = board[0].size

  trie = Trie.new
  words.each {|word| trie.insert(word) }

  cur_word = ''
  ans = []
  visited = Set.new

  dfs = lambda do |x,y, node|
    return if [x<0, y<0, x==m, y==n].any? || visited.include?([x,y])

    c = board[x][y]
    return if node[c].nil?

    cur_word << c
    visited << [x,y]
    ans << cur_word.dup if node[c].is_end

    dfs.call(x-1, y, node[c])
    dfs.call(x+1, y, node[c])
    dfs.call(x, y-1, node[c])
    dfs.call(x, y+1, node[c])

    visited.delete([x,y])
    cur_word.chop!
  end

  (0...m).each do |i|
    (0...n).each do |j|
      dfs.call(i,j, trie)
    end
  end
  ans.uniq
end

class Trie
  attr_accessor :is_end
  def initialize
    @hash = {}
    @is_end = false
  end

  def []=(key,value)
    @hash[key]=value
  end
  def [](key)
    @hash[key]
  end

  def insert(word)
    node = self
    word.each_char do |c|
      node[c] ||= Trie.new
      node = node[c]
    end
    node.is_end = true
  end

  def starts_with(word)
    node = self
    word.each_char do |c|
      return false if node[c].nil?
      node = node[c]
    end
    true
  end

  def has(word)
    node = self
    word.each_char do |c|
      return false if node[c].nil?
      node = node[c]
    end
    node.is_end
  end
end

# @lc code=end

find_words([["a","a"]], ["aaa"])
